pkg/option/config.go

```diff
@@ -898,6 +911,76 @@ func (c *DaemonConfig) Validate() error {
 	return nil
 }
 
+// ReadDirConfig reads the given directory and returns a map that maps the
+// filename to the contents of that file.
+func ReadDirConfig(dirName string) (map[string]interface{}, error) {
+	m := map[string]interface{}{}
+	fi, err := ioutil.ReadDir(dirName)
+	if err != nil && !os.IsNotExist(err) {
+		return nil, fmt.Errorf("unable to read configuration directory: %s", err)
+	}
+	for _, f := range fi {
+		if f.Mode().IsDir() {
+			continue
+		}
+		fName := filepath.Join(dirName, f.Name())
+
+		// the file can still be a symlink to a directory
+		if f.Mode()&os.ModeSymlink == 0 {
+			absFileName, err := filepath.EvalSymlinks(fName)
+			if err != nil {
+				log.Warnf("Unable to read configuration file %q: %s", absFileName, err)
+				continue
+			}
+			fName = absFileName
+		}
+
+		f, err = os.Stat(fName)
+		if err != nil {
+			log.Warnf("Unable to read configuration file %q: %s", fName, err)
+			continue
+		}
+		if f.Mode().IsDir() {
+			continue
+		}
+
+		b, err := ioutil.ReadFile(fName)
+		if err != nil {
+			log.Warnf("Unable to read configuration file %q: %s", fName, err)
+			continue
+		}
+		m[f.Name()] = string(bytes.TrimSpace(b))
+	}
+	return m, nil
+}
+
+// MergeConfig merges the given configuration map with viper's configuration.
+func MergeConfig(m map[string]interface{}) error {
+	err := viper.MergeConfigMap(m)
+	if err != nil {
+		return fmt.Errorf("unable to read merge directory configuration: %s", err)
+	}
+	return nil
+}
+
+// ReplaceDeprecatedFields replaces the deprecated options set with the new set
+// of options that overwrite the deprecated ones.
+func ReplaceDeprecatedFields(m map[string]interface{}) {
+	deprecatedFields := map[string]string{
+		"monitor-aggregation-level":   "monitor-aggregation",
+		"ct-global-max-entries-tcp":   "bpf-ct-global-tcp-max",
+		"ct-global-max-entries-other": "bpf-ct-global-any-max",
+		"legacy-host-allows-world":    "k8s-legacy-host-allows-world",
+	}
```

Comment:
What is the plan around these deprecated options?

`legacy-host-allows-world` is release-noted in v1.4 as deprecated and can be removed in v1.5 (https://github.com/cilium/cilium/issues/6319).

For the other three, it looks like this PR causes the options to be renamed from a user point-of-view. Do we intend to force users to change these in their configmap upon upgrade to v1.5 or will we support these option names for the forseeable future? I think it's important to determine this and describe the intended usage of this function, expected timeline for how long they will be used, etc. in the code as part of the function comment. Given that we need to support these for the upgrade case, and the code is already here, it seems fine to me to mark these three particular options as "we don't intend to get rid of them ever", as it would just introduce unnecessary churn in user configs with no tangible gain.

Reply:
@joestringer An old configmap will always work as long these options stay in this `deprecatedFields` map. I've updated the comment in this function, let me know if it helps.